bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java

```diff
@@ -234,48 +245,116 @@ private TomlNode getTomlNode(TomlNode tomlValue, String variableName, Type type)
+        Optional<Toml> table;
+        String moduleKey = getModuleKey(module);
+        if (hasModuleAmbiguity) {
+            table = baseToml.getTable(moduleKey);
+            if (table.isPresent()) {
+                return table.get().rootNode();
+            }
+            throw new TomlConfigException(String.format(CONTAINS_MODULE_AMBIGUITY, moduleName, moduleKey),
+                        baseToml.rootNode());
+        }
+        table = baseToml.getTable(moduleName);
+        if (table.isEmpty()) {
+            table = baseToml.getTable(moduleKey);
+            if (table.isEmpty() && hasRequired && !invalidRequiredModuleSet.contains(module.toString())) {
+                throwInvalidSubModuleError(baseToml, module);
+            }
         }
-        return extractModuleNode(tomlNode, moduleName, moduleName, hasRequired);
+        return table.map(Toml::rootNode).orElse(null);
     }
 
-    private TomlTableNode validateAndGetModuleStructure(TomlTableNode tomlNode, String key, String moduleName) {
-        TomlNode retrievedNode = tomlNode.entries().get(key);
-        if (retrievedNode != null && retrievedNode.kind() != TomlType.TABLE) {
-            throw new TomlConfigException(String.format(INVALID_MODULE_STRUCTURE, moduleName, moduleName),
-                                          retrievedNode);
+    private void throwInvalidSubModuleError(Toml toml, Module module) {
+        String moduleName = module.getName();
+        TomlNode errorNode = toml.rootNode();
+        Optional<TomlValueNode> valueNode = toml.get(moduleName);
+        List<Toml> tomlTables = toml.getTables(moduleName);
+        if (valueNode.isEmpty()) {
+            valueNode =  toml.get(getModuleKey(module));
+        }
+        if (tomlTables.isEmpty()) {
+            tomlTables =  toml.getTables(getModuleKey(module));
+        }
+        if (valueNode.isPresent()) {
+            errorNode = valueNode.get();
+        } else if (!tomlTables.isEmpty()) {
+            errorNode = tomlTables.get(0).rootNode();
+        } else {
+            Optional<Toml> tomlValueNode = toml.getTable(moduleName.replaceFirst(rootModule.getName() + ".", ""));
+            errorNode = tomlValueNode.isPresent() ? tomlValueNode.get().rootNode() : errorNode;
         }
-        return (TomlTableNode) retrievedNode;
+        invalidRequiredModuleSet.add(module.toString());
+        throw new TomlConfigException(String.format(INVALID_MODULE_STRUCTURE, moduleName, moduleName), errorNode);
     }
 
-    private TomlTableNode extractModuleNode(TomlTableNode orgNode, String moduleName, String fullModuleName,
-                                                   boolean hasRequired) {
-        if (orgNode == null) {
-            return null;
+    private TomlTableNode getRootModuleNode(Toml baseToml) {
+        String moduleName = rootModule.getName();
+        String moduleKey = getModuleKey(rootModule);
+        Optional<Toml> table = baseToml.getTable(moduleKey);
```

Comment:
We need to support following case too.

abc =10
[root_module]
def = 10

so both abc and def variables should get value

Reply:
created issue #29989